### PR TITLE
Feature kraken brokerage setup

### PR DESCRIPTION
**Pull request by kotykhin**
#### Description
Kraken initial setup

#### Related Issue
Closes https://github.com/QuantConnect/Lean/issues/2003

#### Motivation and Context
Kraken reqiured info for research and backtesting

#### Requires Documentation Change
No documentation changes require

#### How Has This Been Tested?
existing tests pass

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Orders/KrakenOrderProperties.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace QuantConnect.Orders
+{
+    public class KrakenOrderProperties : OrderProperties
+    {
+        /// <summary>
+        /// Comma delimited list of order flags. viqc = volume in quote currency (not currently available), fcib = prefer fee in base currency, fciq = prefer fee in quote currency,
+        /// nompp = no market price protection, post = post only order (available when ordertype = limit)
+        /// </summary>
+        public string Oflags { get; set; }
```
Comment (Martin-Molinero) [inline Common/Orders/KrakenOrderProperties.cs:28]: I think we should split out the different order settings as their own properties the users can enable/disable/set, `Comma delimited list of order flags` is going to get error prone and messy. Suggest reviewing existing order properties for other brokerages

File: Common/Orders/Fees/KrakenFeeModel.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        /// Fiats and stablecoins list that have own fee.
+        /// </summary>
+        public readonly List<string> FxStablecoinList = new() {"CAD", "EUR", "GBP", "JPY", "USD", "USDT", "DAI", "USDC"};
+ 
+        /// <summary>
+        /// Get the fee for this order in Quote
+        /// </summary>
+        /// <param name="parameters">A <see cref="OrderFeeParameters"/> object
+        /// containing the security and order</param>
+        /// <returns>The cost of the order USD</returns>
```
Comment (Martin-Molinero) [inline Common/Orders/Fees/KrakenFeeModel.cs:57]: nit - `The cost of the order USD` think this is off?

File: Common/Orders/Fees/KrakenFeeModel.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            if (FxStablecoinList.Any(i => security.Symbol.Value.StartsWith(i)))
+            {
+                fee = Tier1FxFee;
+            }
+            
+            return new OrderFee(new CashAmount(
+                unitPrice * order.AbsoluteQuantity * fee,
+                security.QuoteCurrency.Symbol));
```
Comment (Martin-Molinero) [inline Common/Orders/Fees/KrakenFeeModel.cs:88]: Let's add some unit tests for this class 👍  see examples at https://github.com/QuantConnect/Lean/tree/master/Tests/Common/Orders/Fees

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            {"EOS", 3}, // eth available
+            {"DASH", 3}, 
+            {"TRX", 3}, // eth available
+            {"LINK", 3}, // eth available
+            {"USDC", 3}, // only with fiats
+        };
+
+        private readonly List<string> _fiatsAvailableMargin = new() {"USD", "EUR"};
+        private readonly List<string> _onlyFiatsAvailableMargin = new() {"BTC", "USDT", "USDC"};
+        private readonly List<string> _ethAvailableMargin = new() {"REP", "XTZ", "ADA", "EOS", "TRX", "LINK" };
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:65]: nit - let's keep private properties and fields at the top and private methods at the bottom 👍

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        /// <returns>The buying power model for this brokerage/security</returns>
+        public override IBuyingPowerModel GetBuyingPowerModel(Security security)
+        {
+            if (AccountType == AccountType.Margin)
+            {
+                foreach (var lev in CoinLeverage)
+                {
+                    if (security.Symbol.Value.StartsWith(lev.Key))
+                    {
+                        return new SecurityMarginModel(lev.Value);
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:124]: > return new SecurityMarginModel(lev.Value);

To be consistent should maybe call `GetLeverage()` and pass in the returned value?

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            // first check whether this security support margin only with fiats.
+            foreach (var coin in _onlyFiatsAvailableMargin.Where(coin => security.Symbol.Value.StartsWith(coin)).Where(coin => _fiatsAvailableMargin.Any(rightFiat => security.Symbol.Value.EndsWith(rightFiat))))
+            {
+                return CoinLeverage[coin];
+            }
+
+            List<string> extendedCoinArray = new() {"BTC", "ETH"};
+            extendedCoinArray.AddRange(_fiatsAvailableMargin);
+            // Then check whether this security support margin with ETH.
+            foreach (var coin in _ethAvailableMargin.Where(coin => security.Symbol.Value.StartsWith(coin)).Where(coin => extendedCoinArray.Any(rightFiat => security.Symbol.Value.EndsWith(rightFiat))))
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:153]: In general shouldn't use `symbol.Value` but `Symbol.ID.Symbol`, the value value can change and get remapped by lean or be an alias in some cases

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            {
+                return CoinLeverage[coin];
+            }
+
+            extendedCoinArray.Remove("ETH");
+            // At the end check all others.
+            foreach (var coin in CoinLeverage.Keys.Where(coin => security.Symbol.Value.StartsWith(coin)).Where(coin => extendedCoinArray.Any(rightFiat => security.Symbol.Value.EndsWith(rightFiat))))
+            {
+                return CoinLeverage[coin];
+            }
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:163]: Cool! let's please add a nice test collection for all these different returned leverage cases 👍

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Awesome! Leaving a few requests

File: Data/symbol-properties/symbol-properties-database.csv [hunk 8]
```diff
@@ -2306,3 +2306,391 @@ binance,ZRXBTC,crypto,ZRXBTC,BTC,1,0.00000001,1.00000000,ZRXBTC
 binance,ZRXBUSD,crypto,ZRXBUSD,BUSD,1,0.00010000,0.01000000,ZRXBUSD
 binance,ZRXETH,crypto,ZRXETH,ETH,1,0.00000001,1.00000000,ZRXETH
 binance,ZRXUSDT,crypto,ZRXUSDT,USDT,1,0.00010000,0.01000000,ZRXUSDT
+
+# kraken has different naming for ws and rest, so description is ws name
```
[the remark this answers is not in the document]
Comment (kotykhin) [reply Data/symbol-properties/symbol-properties-database.csv:2310]: Kraken has weird mapping, e.g lean symbol BTCUSD in Kraken rest is XXBTZUSD and in ws is XBT/USD. We can't choose 1 symbol and map it to another - coins are different, so each name's important. \
Clean solution, I think, to add a new field to SymbolProperties, but I decided not to edit source code and come up with this workaround

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Data/symbol-properties/symbol-properties-database.csv:2310]: > Kraken rest is XXBTZUSD and in ws is XBT/USD

oh for what is kraken rest used for, to subscribe to the data? Crazy to subscribe with one ticker value and get data with another ticker value

[the remark this answers is not in the document]
Comment (kotykhin) [reply Data/symbol-properties/symbol-properties-database.csv:2310]: Kraken rest is used for all rest requests we're doing (e.g history, positions etc)\
Ws symbol used for websockets needs - subscribing to channels, recieving data and even placing orders. I mean,
If you want to place order of BTCUSD by rest request - you need to send  XXBTZUSD order, but if you want to place same order via websocket, you need to send XBT/USD. Different naming but the same result

File: Tests/Brokerages/Kraken/KrakenBrokerageModelTests.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                new Cash(Currencies.USD, 0, 1m),
+                SymbolProperties.GetDefault(Currencies.USD),
+                ErrorCurrencyConverter.Instance,
+                RegisteredSecurityDataTypesProvider.Null,
+                new SecurityCache()
+            );
+
+            var leverage = _brokerageModel.GetLeverage(security);
+            
+            Assert.True(leverage == expectedLeverage, "Expected leverage doesn't match with returned");
```
Comment (Martin-Molinero) [inline Tests/Brokerages/Kraken/KrakenBrokerageModelTests.cs:70]: nit - `Assert.True` should be `Assert.AreEqual` here?

File: Common/Orders/KrakenOrderProperties.cs (hunk 1 above)
Comment (Martin-Molinero) [reply Common/Orders/KrakenOrderProperties.cs:28]: `Oflags` can now be removed?

nit - Same for `using System;` really

File: Common/Orders/KrakenOrderProperties.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        
+        /// <summary>
+        /// https://support.kraken.com/hc/en-us/articles/201648183-Market-Price-Protection
+        /// </summary>
+        public bool NoMarketPriceProtection { get; set; }
+
+        /// <summary>
+        /// Conditional close orders are triggered by execution of the primary order in the same quantity and opposite direction. Ordertypes can be the same with primary order.
+        /// </summary>
+        public Order ConditionalOrder { get; set; } = null;
```
Comment (Martin-Molinero) [inline Common/Orders/KrakenOrderProperties.cs:53]: nit - `null` is already the default here?

**Review by Martin-Molinero — APPROVED** [review]
Cool! last minor comments, looks good 👍

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 11]
```diff
@@ -95,6 +95,20 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
...
+        
+        /// <summary>
+        /// Kraken does no support update of orders
+        /// </summary>
+        /// <param name="security"></param>
+        /// <param name="order"></param>
+        /// <param name="request"></param>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        public override bool CanUpdateOrder(Security security, Order order, UpdateOrderRequest request, out BrokerageMessageEvent message)
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:107]: nit - missing method docs

**Review by Martin-Molinero — APPROVED** [review]

File: Common/Brokerages/KrakenBrokerageModel.cs (hunk 11 above)
Comment (Martin-Molinero) [reply Common/Brokerages/KrakenBrokerageModel.cs:107]: *needs a rebase too 👍

Comment (kotykhin) [reply Common/Brokerages/KrakenBrokerageModel.cs:107]: Can you explain please what do you mean by 'rebase method docs'?

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**